File: test-suite/handwritten-src/js/ClientInterfaceTest.js
class ClientInterfaceImpl {
    getRecord(id, utf8String, misc) {
        if (utf8String != "Non-ASCII /\0 非 ASCII 字符" && utf8String != "Hello World!") {
            throw new Error("Unexpected string. Check UTF-8?");
        }
        return {recordId: id,
                content: utf8String,
                misc: misc};
    }
    identifierCheck(data, r, jret) {
        return 0.0;
    }
    returnStr() {
        return "test";
    }
    methTakingInterface(i) {
        if (i != null) {return i.returnStr();} else {return "";}
    }
    methTakingOptionalInterface(i) {
        if (i != null) {return i.returnStr();} else {return "";}
    }
}

class ClientInterfaceTest {
    constructor(module) {
        this.module = module;
    }
    setUp() {
        this.jsClientInterface = new ClientInterfaceImpl();
    }
    testClientReturn() {
        this.module.testsuite.TestHelpers.checkClientInterfaceAscii(this.jsClientInterface);
    }

    testClientReturnUTF8() {
        this.module.testsuite.TestHelpers.checkClientInterfaceNonascii(this.jsClientInterface);
    }

    testClientInterfaceArgs() {
        this.module.testsuite.TestHelpers.checkClientInterfaceArgs(this.jsClientInterface);
    }

    testReverseClientInterfaceArgs() {
        var i = this.module.testsuite.ReverseClientInterface.create();
        assertEq(i.methTakingInterface(i), "test");
        assertEq(i.methTakingOptionalInterface(i), "test");
    }
}

allTests.push(ClientInterfaceTest);

File: test-suite/handwritten-src/swift/ClientInterfaceTest.swift
import XCTest
@testable import TestSuite


struct ClientInterfaceError: Error {
    var msg: String
}

class ClientInterfaceImpl: ClientInterface {
    func getRecord(_ recordId: Int64, utf8string: String, misc: Optional<String>) throws -> TestSuite.ClientReturnedRecord {
        if (utf8string != "Non-ASCII /\0 非 ASCII 字符" && utf8string != "Hello World!") {
            throw ClientInterfaceError(msg: "Unexpected string. Check UTF-8?")
        }
        return ClientReturnedRecord(recordId:recordId, content:utf8string, misc:misc)
    }
    func identifierCheck(_ data: Data, r: Int32, jret: Int64) throws -> Double {
        return 0.0
    }
    func returnStr() throws -> String {
        return "test"
    }
    func methTakingInterface(_ i: TestSuite.ClientInterface) throws -> String {
        return try i.returnStr()
    }
    func methTakingOptionalInterface(_ i: Optional<TestSuite.ClientInterface>) throws -> String {
        if (i != nil) {
            return try i!.returnStr()
        } else {
            return ""
        }
    }
}

final class ClientInterfaceTest: XCTestCase {
    var sClientInterface: ClientInterface? = nil

    override func setUp() {
        sClientInterface = ClientInterfaceImpl()
    }

    func testClientReturn() throws {
        try TestHelpers_statics.checkClientInterfaceAscii(sClientInterface!)
    }

    func testClientReturnUTF8() throws {
        try TestHelpers_statics.checkClientInterfaceNonascii(sClientInterface!)
    }

    func testClientInterfaceArgs() throws {
        try TestHelpers_statics.checkClientInterfaceArgs(sClientInterface!)
    }

    func testReverseClientInterfaceArgs() throws {
        let i = try ReverseClientInterface_statics.create()
        XCTAssertEqual(try i.methTakingInterface(i), "test")
        XCTAssertEqual(try i.methTakingOptionalInterface(i), "test")
    }
}
